// fw-object-validator/src/messages.js
module.exports = {
    required: 'The Field {param} is required.',
    string: `The parameter '{param}' have to be a string.`,
    stringMin: `The parameter '{param}' must be at least {expected} characters long.`,
    stringMax: `The parameter '{param}' must be at least {expected} characters long.`,
    stringEqual: `The parameter '{param}' have to equal {expected}.`,
    stringContain: `The parameter '{param}' have to contain {expected}.`,

    boolean: `The parameter '{param}' have to be a boolean value.`,
    array: `The parameter '{param}' have to be an array.`,
    date: `The parameter '{param}' is not a valid date.`,
    email: `The value of parameter '{param}' is not a valid email address.`,
    forbidden: `The parameter '{param}' is forbidden.`,
    number: `The parameter '{param}' have to be a number.`,
    numberMin: `The type of parameter '{param}' have to be boolean.`,
    numberMax: `The type of parameter '{param}' have to be boolean.`,
    numberInteger: `The type of parameter '{param}' have to be boolean.`,
    numberPositive: ``,
    numberNegative: ``,
    object: ``,
    url: ``
}
